contracts/entry-point/src/execute.rs

```diff
@@ -182,8 +182,37 @@ pub fn execute_swap_and_action(
     let exact_out = match &user_swap {
         Swap::SwapExactAssetIn(_) => false,
         Swap::SwapExactAssetOut(_) => true,
+        Swap::SmartSwapExactAssetIn(_) => false,
     };
 
+    if let Swap::SmartSwapExactAssetIn(smart_swap) = &mut user_swap {
+        if smart_swap.routes.is_empty() {
+            return Err(ContractError::Skip(skip::error::SkipError::RoutesEmpty));
+        }
+
+        match smart_swap.amount().cmp(&remaining_asset.amount()) {
+            std::cmp::Ordering::Equal => {}
+            std::cmp::Ordering::Less => {
+                let diff = remaining_asset.amount().checked_sub(smart_swap.amount())?;
+
+                // If the total swap in amount is less than remaining asset,
+                // adjust the routes to match the remaining asset amount
+                let largest_route_idx = smart_swap.largest_route_index().unwrap();
```

Comment:
nit: Let's use the "?" , trying to avoid unwrap() in code, should result in same thing and never be triggered so no biggy but want to avoid unwraps as much as possible as a general rule